Add the first iteration of the `util.hashing` package.

axel22 opened the pull request:
Next step is to move `MurmurHash3` to `util.hashing`,
make the `class` private and retain a public companion
`object`, and put the `MurmurHash3.Hashing` implementations
for various types in the companion.

Review by @jsuereth.

Commit:
Add the first iteration of the `util.hashing` package.

Next step is to move `MurmurHash3` to `util.hashing`,
make the `class` private and retain a public companion
`object`, and put the `MurmurHash3.Hashing` implementations
for various types in the companion.

Review by @jsuereth.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/231/

jsuereth commented on src/library/scala/util/hashing/ByteswapHashing.scala:
```diff
@@ -0,0 +1,20 @@
+
+
+
+final class ByteswapHashing[T] extends Hashing[T] {
```
Should this chain into another Hashing instance?  I can see that being useful potentially.

axel22 replied:
It seems useful, yes.
Maybe we should have a separate class in that case which composes 2 `Hashing`s together.

axel22 replied:
On a second thought, that would require changing the `Hashing` interface.

I would have a separate class then for the `Byteswap`, to avoid indirections if only the regular `Byteswap` is to be used.

jsuereth commented on src/library/scala/util/hashing/package.scala:
```diff
@@ -0,0 +1,35 @@
+
+
+
+package object hashing {
```
Looks good.  Mind if I add a CityHash implementation in here?

axel22 replied:
Not at all.

jsuereth commented:
Hmm... am I missing the `def hashCode` -> `hashOf` move?  i.e. is a class missing in the commit?

Other than that, LGTM where G = great.

axel22 commented:
Some guy complained about it, saying that an overload might be confusing.
Seems clearer to me, besides we also use `equiv` in the `Equiv` instead of `equals`.

axel22 commented:
The move was in the `Hashing` trait.

jsuereth commented on src/library/scala/util/hashing/Hashing.scala:
```diff
@@ -22,7 +22,7 @@ package scala.util.hashing
 trait Hashing[T] extends Serializable {
   
-  def hashCode(x: T): Int
+  def hashOf(x: T): Int
```
AH.. I hate to be annoying, but why not just call it 'hash' if we're going to change it.   Either one is fine really.

axel22 replied:
No problem.

Commit:
Move `MurmurHash3` to `util.hashing`.

Add a method which composes `ByteswapHashing` with some other hashing.

Rename `hashOf` to `hash`.

axel22 commented:
Updated the commit range. We might have to close to restart bot.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/231/

axel22 commented:
Lets close this and I'll submit a new pull.

paulp commented:
Do I understand correctly that the name is changing? Based on my experience with peoples' overloads of toString, you do not, not, not, not, not want a method with a signature of "def hashCode(x: T): Int". You never want to overload anything defined in java.lang.Object, and if you can help it, never overload across inheritance.  The unfortunate side effects for type inference and etc are never-ending.

jsuereth commented:
Yeah, we're changing `def hashCode(x: T): Int` into `def hash(x: T): Int` or `def hashFor(x: T): Int`

axel22 commented:
In my last commit, it's `def hash(x: T): Int`.
